093.mediawiki

```diff
@@ -0,0 +1,329 @@
+We describe each of these complex structures separately.
+
+====CGrapheneBlock====
+''CGrapheneBlock'' is the fundamental data structure for Graphene block propagation. Assuming no unrecoverable failures arise, and combined with any missing transactions recovered from the sender, this data structure contains all information necessary to reconstruct the full block. It is sent as part of the '''grblk''' message.
+
+{|
+ |'''Field Name'''||'''Type'''||'''Size'''||'''Purpose'''
+ |-
+ |header||Block header||80 bytes||The header of the block being provided, encoded as per the header portion of the BLOCK message (NOT including the nTx field of the HEADERS message)
+ |-
+ |vAdditionalTxs||<nowiki>vector<CTransaction></nowiki>||variable||Transactions that the receiver is probably missing. Standard transaction Bitcoin serialization is used.
+ |-
+ |nBlockTxs||uint64_t||8 bytes||Number of Transactions in the block
+ |-
+ |grapheneSet||CGrapheneSet||variable||Encapsulates Graphene set reconciliation logic
+ |}
+
+The vAdditionalTxs field should include at least the coinbase transaction, since it is not possible for it to be in the receiver's mempool.
+
+====CGrapheneSet====
+''CGrapheneSet'' contains all information critical to the Graphene set reconciliation process, absent any block-specific details. It is the only non-standard data structure in ''CGrapheneBlock''.
+
+{|
+ |'''Field Name'''||'''Type'''||'''Size'''||'''Purpose'''
+ |-
+ |ordered||uint8_t||1 byte||1 if order is important in the set and 0 otherwise; other values are reserved for future use
+ |-
+ |nReceiverUniverseItems||uint64_t||8 bytes||The number of transactions in the receiver's mempool
+ |-
+ |encodedRank||<nowiki>vector<unsigned char></nowiki>||ordered * ceil(n log(n))/8 bytes||Order information for items in set
+ |-
+ |setFilter||CBloomFilter||variable||Bloom filter containing items in set
+ |-
+ |setIblt||CIblt||variable||IBLT containing items in set
+ |}
+
+====CBloomFilter====
+''CBloomFilter'' is Bitcoin's standard Bloom filter implementation with one adjustment, described below. It is used as part of the set reconciliation process in ''CGrapheneSet'' to filter out transactions from ''T'' that do not belong to the block.
+
+{|
+ |'''Field Name'''||'''Type'''||'''Size'''||'''Purpose'''
+ |-
+ |vData||<nowiki>vector<unsigned char></nowiki>||variable||Bit array for filter
+ |-
+ |isFull||bool||1 byte||True if every bit in vData is set to 1
+ |-
+ |isEmpty||bool||1 byte||True if every bit in vData is set to 0
```

Comment:
isFull and isEmpty can be merged into a single 1-byte field (state, perhaps?) where simple boolean flags may be used. 0x1 for isEmpty and 0x2 for isFull, would be my suggestion.

Reply:
I just like to note here that this is stemming from the current `CBloomFilter` serialization that we all inherited from Core.

Reply:
Like @awemany stated, the CBloomFilter data structure is currently based on the legacy class from Core. I think there is a legitimate discussion that could be had about whether or not a separate, dedicated Bloom filter class should be used for graphene. Probably such a discussion is above me though :-).